include/up-cpp/transport/UTransport.h

```diff
@@ -0,0 +1,206 @@
+/// @brief Abstract base class for all transport implementations.
+///
+/// All implementations derived from this class must meet the following
+/// requirements:
+///
+/// * [MUST] An implementation MUST support multiple simultaneous instantiatons
+///          from within the same application.
+///
+/// * [MAY]  An implimentation MAY require that each new instance within an
+///          application have a unique configuration from existing instances.
+///
+/// * [MUST] An implementation MUST allow for multiple users of the same
+///          instance.
+///
+/// * [MUST] An implementation MUST be thread-safe.
+///
+/// * [MUST] Throw an exception if the transport fails to initialize or the
+///          configuration is invalid.
+class UTransport {
+	/// @brief Connection interface used for self-terminating listener
+	/// registrations
+	using CallbackConnection =
+	    utils::callbacks::Connection<void, const v1::UMessage&>;
+
+public:
+	/// @brief Constructor
+	///
+	/// @param Default Authority and Entity (as a UUri) for clients using this
+	///        transport instance.
+	///
+	/// @throws uprotocol::datamodel::validator::uuri::InvalidUUri
+	///         if the provided UUri contains a resource, or is missing an
+	///         authority or entity.
+	explicit UTransport(const v1::UUri&);
+
+	/// @brief Send a message.
+	///
+	/// Must be implemented by the transport client library.
+	///
+	/// @param message UMessage to be sent. Contains all the information
+	///                needed to send the message as UAttributes and
+	///                UPayload
+	///
+	/// @returns * OKSTATUS if the payload has been successfully
+	///            sent (ACK'ed)
+	///          * FAILSTATUS with the appropriate failure.
+	[[nodiscard]] virtual v1::UStatus send(const v1::UMessage& message) = 0;
+
+	/// @brief Callback function (void(const UMessage&))
+	using ListenCallback = typename CallbackConnection::Callback;
+
+	/// @brief Handle representing the callback connection.
+	///
+	/// These handles will automatically disconnect the callback when
+	/// destroyed or when .reset() is called. They can be cast to bool
+	/// to check if they are connected.
+	using ListenHandle = typename CallbackConnection::Handle;
+
+	/// @brief Register listener to be called when UMessage is received
+	///        for the given URI.
+	///
+	/// @param uri Resolved UUri for where messages are expected to
+	///            arrived via the underlying transport technology.
+	/// @param listener Callback to be called when a message is received
+	///                 at the given URI. The UMessage will be provided
+	///                 as a parameter when the callback is called.
+	///
+	/// @returns * OKSTATUS and a connected ListenHandle if the listener
+	///            was registered successfully.
+	///          * FAILSTATUS with the appropriate failure and an
+	///            unconnected ListenHandle otherwise.
+	[[nodiscard]] std::tuple<v1::UStatus, ListenHandle> registerListener(
```

Comment:
Not sure I understand the public vs protected APIs and why they are different

Reply:
The protected ones are pure virtual and must be implemented by the derived transports (e.g. ZenohUTransport).

The public ones are not virtual at all. They will handle the creation of the `CallbackConnection::ConnectedPair`, handing the `Callable` half to the protected `registerListener` and the `Handle` half to the caller of this method.